src/haswell/avx2_convert_utf32_to_utf16.cpp

```diff
@@ -1,76 +1,130 @@
+  const uint8_t mask_lo = mask_2x & 0xf;
+  const uint8_t mask_hi = mask_2x >> 4;
+
+  const auto t0 = in - uint32_t(0x00010000);
+  const auto hi = t0.shr<10>() & uint32_t(0x000003ff);
+  const auto lo = t0.shl<16>() & uint32_t(0x03ff0000);
+  const auto surrogates = (lo | hi) | uint32_t(0xdc00d800);
+
+  const auto merged = as_vector_u8(select(non_surrogate_mask, in, surrogates));
+
+  // Note: we're getting back to the intrinsics world, as there's
+  //       no way to refer to the westmere implementation (we enclose
+  //       implementations in anonymous namespaces).
+  const auto lo_half = _mm256_extractf128_si256(merged.value, 0);
+  const auto hi_half = _mm256_extractf128_si256(merged.value, 1);
+
+  const auto shuffle_lo = _mm_loadu_si128(reinterpret_cast<const __m128i *>(
+      (byte_order == endianness::LITTLE)
+          ? tables::utf32_to_utf16::pack_utf32_to_utf16le[mask_lo]
+          : tables::utf32_to_utf16::pack_utf32_to_utf16be[mask_lo]));
+
+  const auto shuffle_hi = _mm_loadu_si128(reinterpret_cast<const __m128i *>(
+      (byte_order == endianness::LITTLE)
+          ? tables::utf32_to_utf16::pack_utf32_to_utf16le[mask_hi]
+          : tables::utf32_to_utf16::pack_utf32_to_utf16be[mask_hi]));
+
+  const size_t u16count_lo = (4 + count_ones(mask_lo));
+  const size_t u16count_hi = (4 + count_ones(mask_hi));
+
+  const auto compressed_lo = _mm_shuffle_epi8(lo_half, shuffle_lo);
+  const auto compressed_hi = _mm_shuffle_epi8(hi_half, shuffle_hi);
+
+  return {u16count_lo, compressed_lo, u16count_hi, compressed_hi};
+}
+
+// Function `invalid_utf32` returns a bytemask pointing invalid UTF-32
+// characters.
+simdutf_really_inline simd32<bool> invalid_utf32(const __m256i x) {
+  using vector_u32 = simd32<uint32_t>;
+
+  const auto in = vector_u32(x);
+
+  const auto standardmax = vector_u32::splat(0x10ffff);
+
+  const auto too_large = in > standardmax;
+  const auto surrogate = (in & uint32_t(0xfffff800)) == uint32_t(0xd800);
+
+  return too_large | surrogate;
+}
+
 template <endianness big_endian>
 std::pair<const char32_t *, char16_t *>
 avx2_convert_utf32_to_utf16(const char32_t *buf, size_t len,
                             char16_t *utf16_output) {
   const char32_t *end = buf + len;
 
-  const size_t safety_margin =
-      12; // to avoid overruns, see issue
-          // https://github.com/simdutf/simdutf/issues/92
+  // to avoid overruns, see issue https://github.com/simdutf/simdutf/issues/92
+  const size_t safety_margin = 12;
+
   __m256i forbidden_bytemask = _mm256_setzero_si256();
 
   const __m256i v_ffff0000 = _mm256_set1_epi32((int32_t)0xffff0000);
   const __m256i v_f800 = _mm256_set1_epi32((uint32_t)0xf800);
   const __m256i v_d800 = _mm256_set1_epi32((uint32_t)0xd800);
 
-  while (end - buf >= std::ptrdiff_t(8 + safety_margin)) {
-    const __m256i in = _mm256_loadu_si256((__m256i *)buf);
+  while (end - buf >= std::ptrdiff_t(16 + safety_margin)) {
```

Comment:
Sorry for the basic question - trying to understand why we need `(16 + 12) * 4` which is 128 bytes of input available for this loop.
As per my understanding the loop reads 32 bytes (8 words) and this could expand to 16 utf16 halfwords. Considering that the second half of vector can write out of bounds upto 8 bytes (4 halfwords), wouldn't it be enough for `safety_margin` to be 4 to guarantee further 8 bytes of write later by scalar code? So wouldn't it be sufficient for the loop check to be something like `end - buf >= std::ptrdiff_t(12)`?

Reply:
It's a mistake, it should still be `8 + safety_margin`. I tried unrolling the loop twice, and forgot to revert this change.

Reply:
Should `safety_margin` still be `12` (48 bytes)? What is the reason for this large margin?

Reply:
Hm, that's a good question. I need to recheck this.